core/divelist.c

```diff
@@ -367,94 +438,128 @@ int get_divesite_idx(struct dive_site *ds)
-	when = dive->when;
-	i = divenr;
-	if (i < 0) {
-		i = dive_table.nr - 1;
-		while (i >= 0 && get_dive(i)->when > when)
-			--i;
+	i = divenr >= 0 ? divenr : dive_table.nr;
+	if (i >= 0 && i < dive_table.nr) printf("\n\n*** Init deco for dive #%d %d\n", i, get_dive(i)->number);
+	else printf("\n\n*** Init deco for dive #%d\n", i);
+	/* Look at next dive in dive list table and correct i when needed */
+	while (i < dive_table.nr - 1) {
+		struct dive *pdive = get_dive(i);
+		if (!pdive || pdive->when > dive->when)
+			break;
 		i++;
 	}
+	/* Look at previous dive in dive list table and correct i when needed */
+	while (i > 0) {
+		struct dive *pdive = get_dive(i - 1);
+		if (!pdive || pdive->when < dive->when)
+			break;
+		i--;
+	}
+	printf("Dive number corrected to #%d\n", i);
+	last_starttime = dive->when;
+	/* Walk backwards to check previous dives - how far do we need to go back? */
 	while (i--) {
+		if (i == divenr && i > 0) i--;
+		printf("Check if dive #%d %d has to be considered as prev dive: ", i, get_dive(i)->number);
 		struct dive *pdive = get_dive(i);
 		/* we don't want to mix dives from different trips as we keep looking
 		 * for how far back we need to go */
-		if (dive->divetrip && pdive->divetrip != dive->divetrip)
+		if (dive->divetrip && pdive->divetrip != dive->divetrip) {
+			printf("No - other dive trip\n"); 
 			continue;
-		if (!pdive || pdive->when >= when || dive_endtime(pdive) + 48 * 60 * 60 < when)
+		}
+		if (!pdive || pdive->when >= dive->when || dive_endtime(pdive) + 48 * 60 * 60 < last_starttime) {
+			printf("No\n");
 			break;
-		/* For simultaneous dives, only consider the first */
-		if (pdive->when == laststart)
-			continue;
-		lasttime = dive_endtime(pdive);
+		}
+		last_starttime = pdive->when;
+		printf("Yes\n");
 	}
-	while (++i < (divenr >= 0 ? divenr : dive_table.nr)) {
+	/* Walk forward an add dives and surface intervals to deco */
+	while (++i < dive_table.nr) {
+		printf("Check if dive #%d %d will be really added to deco calc: ", i, get_dive(i)->number);
 		struct dive *pdive = get_dive(i);
 		/* again skip dives from different trips */
-		if (dive->divetrip && dive->divetrip != pdive->divetrip)
+		if (dive->divetrip && dive->divetrip != pdive->divetrip) {
+			printf("No - other dive trip\n"); 
 			continue;
+		}
 		/* Don't add future dives */
-		if (pdive->when > dive->when)
-			continue;	/* This could be break if the divelist is always sorted */
+		if (pdive->when >= dive->when) {
+			printf("No - future or same dive\n");
+			break;
+		}
+		/* Don't add the copy of the dive itself */
+		if (i == divenr) {
+			printf("No - copy of dive\n");
+			continue;
+		}
+		printf("Yes\n");
+
 		surface_pressure = get_surface_pressure_in_mbar(pdive, true) / 1000.0;
+		/* Is it the first dive we add? */
 		if (!deco_init) {
+			printf("Init deco\n");
 			clear_deco(surface_pressure);
 			deco_init = true;
-#if DECO_CALC_DEBUG & 2
+			printf("Tissues after init:\n");
 			dump_tissues();
-#endif
```

Comment:
we probably don't want to dump the tissues all the time by removing the `#if DECO_CALC_DEBUG & 2`. i'm assuming you did this so that you can monitor the changes without supplying the macro definition to the compiler?

Reply:
Yes, for the final version I will for sure switch off the debug output by default again. I maybe need some hint how to do it: Remove it completely or add the #if DECO_CALC_DEBUG & 2 everywhere again.